File: baekjoon/13023.cpp
#include <bits/stdc++.h>
#define MAX_VERT 2001
#define endl '\n'
using namespace std;

int V, E;
bool flag=false;
vector<int> adj[MAX_VERT];
vector<bool> visited(MAX_VERT, false);

void input() noexcept
{
	int node1, node2;
	cin >> V >> E;

	for (int i = 0; i < E; i++)
	{
		cin >> node1 >> node2;
		adj[node1].push_back(node2);
		adj[node2].push_back(node1);
	}
}

void dfs(int node, int dpt) noexcept
{
	if (dpt == 5)
	{
		flag = true;
		return;
	}
	visited[node] = true;
	for(const auto&ele: adj[node])
	{
		if (!visited[ele])
			dfs(ele, dpt + 1);
	}
	visited[node] = false;
}

void solve() noexcept
{
	input();
	for (int i = 0; i < V; i++)
	{
		if (!flag)
			dfs(i, 1);
		fill(visited.begin(), visited.end(), false);
	}
	if (flag)
		cout << 1;
	else
		cout << 0;
}


int main(void) noexcept
{
	ios::sync_with_stdio(0);
	cin.tie(0); cout.tie(0);

	solve();

	return (0);
}

File: baekjoon/java/13023.java
import java.io.*;
import java.util.*;

public class Main {

    static int V, E;
    static boolean flag = false;
    static List<List<Integer>> adj = new ArrayList<>();
    static boolean[] visited;

    public static void main(String[] args) throws IOException {
        solve();
    }

    static void solve() throws IOException {
        input();
        visited = new boolean[V];

        for (int i = 0; i < V; i++) {
            if (!flag) {
                dfs(i, 1);
            }
            Arrays.fill(visited, false);  // 방문 배열 초기화
        }

        System.out.println(flag ? 1 : 0);
    }

    static void input() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        V = Integer.parseInt(st.nextToken());
        E = Integer.parseInt(st.nextToken());

        for (int i = 0; i < V; i++) {
            adj.add(new ArrayList<>());
        }

        for (int i = 0; i < E; i++) {
            st = new StringTokenizer(br.readLine());
            int node1 = Integer.parseInt(st.nextToken());
            int node2 = Integer.parseInt(st.nextToken());

            adj.get(node1).add(node2);
            adj.get(node2).add(node1);
        }
    }

    static void dfs(int node, int depth) {
        if (depth == 5) {
            flag = true;
            return;
        }

        visited[node] = true;
        for (int neighbor : adj.get(node)) {
            if (!visited[neighbor]) {
                dfs(neighbor, depth + 1);
            }
        }
        visited[node] = false;
    }
}
